design-proposals/template-generation.md

```diff
@@ -0,0 +1,131 @@
+# Overview
+
+This design defines an approach to create Virtual Machine templates starting from an existing VM.
+
+## Motivation
+
+Creating a 'blueprint' of an existing VM to further tweak and use as a basis golden VM template is a common operation, also it is a tool that can be leveraged by UIs to improve the UX of the project.
+
+## Goals
+
+* Create a generic template of an existing VM.
+
+## Non Goals
+
+* Designing a VM-cloning mechanism, which copies disk data or create a snapshot of the live memory of a VM.
+
+## User Stories
+
+* As a cluster user, I want to create a re-usable VM spec to use as the basis for my VMs workloads starting from an existing VM.
+* As a cluster user, I want to create different workloads based on the same VM image by just tweaking some template parameters.
+
+## Repos
+
+The idea is to implement this design by adding a controller in kubevirt/ssp-operator.
+
+# Design
+
+The template generation API introduces a new CRD called the 'VirtualMachineTemplateRequest' object. This CRD when posted to the cluster is an explicit request to create a VM template derived from an existing VM.
+
+The controller watching `VirtualMachineTemplateRequest` objects will see the request and post a template of the request VirtualMachine in the specified format.
+    
+## VirtualMachineTemplateRequest API
+
+The VirtualMachineTemplateRequest specification fields:
+
+* **VMName**: the name of a VirtualMachine living in the cluster which will be the reference from which the template will be created (required field).
+* **VMNamespace**: the namespace of the reference VM. If not specified the VM will be looked up in the namepace in which VirtualMachineTemplateRequest was created (optional field).
+* **TemplateName**: the name of the template object that will be generated, if omitted it will be auto-generated (otional field).
+* **Format**: the format of the output template (required field).
+
+The first implemented format will be the [OpenShift Template(https://docs.openshift.com/container-platform/4.9/openshift_images/using-templates.html#templates-writing_using-templates)] format.
+
+In order to support multiple formats, the controller will have to be extended to generate the correct output for those formats.
+
+## API Examples
+
+Given the following VM
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-ephemeral
+spec:
+  template:
+    domain:
+      devices:
+        disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      resources:
+        requests:
+          memory: 128Mi
+    volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/cirros-container-disk-demo:devel
+      name: containerdisk
+```
+
+After posting this VirtualMachineTemplateRequest object
+
+```yaml
+apiVersion: ssp.kubevirt.io/v1alpha1
+kind: VirtualMachineTemplateRequest
+metadata:
+  name: virtualmachinetemplaterequest-sample
+spec:
+  vmName: vmi-ephemeral
```

Comment:
What about differentiating source and target more:

```
from:
  vmName: vmi-ephemeral
to:
  openshiftTemplate:
    name: vmi-ephemeral-template
``

The benefit of a dedicated struct instead of a format value is, is that the API is "future proof", we can easily add format specific fields in future and keepin backwards compatibility.